book/src/developers/designs/application_passwords.md

```diff
@@ -0,0 +1,215 @@
+cache applications and just compare their member sets to the user sessions
+memberof)
+
+Applications need to have service-account like properties where tokens can be
+generated for them. These are optional since an anonymous bind to kanidm and
+searching under the basedn will continue to work.
+
+Application should have a URL reference to help admins identify where the
+application may be located or accessed.
+
+(NOTE: Future, it could be good to allow customisable instructions for users on
+where to go to use their app password?)
+
+(NOTE: To make it easier to manage tokens, it could be good to have a
+"token-capable" class type to prevent duplication this with service-accounts,
+and allows adding tokens to future types).
+
+### Accounts
+
+The user may wish to have multiple passwords per application. Each password may
+have a label to identify it. For example:
+
+```
+  MAIL
+    iphone: abcd...
+    laptop: bcde...
+  HTTP
+    workstation: cdef...
+```
+
+Person accounts will need a new `Attribute::ApplicationPassword` that stores a
+`ValueSetApplicationPassword`. Each value in the set is a new type to manage these
+secrets and their labeling and the references to the applications.
+
+```
+type ApplicationPasswords = BTreeMap<Uuid, (String, Password)>;
+                                     ^        ^         ^
+                                     |        |         +-> The hashed password
+                                     |        +-> Application password label
+                                     +-> Application password UUID
+
+type ApplicationUuid = Uuid;
+
+struct ValueSetApplicationPassword {
+  map: Map<ApplicationUuid, ApplicationPasswords>
+}
+```
+
+Each value in the set is queried by its UUID. This defines the value as
+`Value::ApplicationPassword(Uuid, ApplicationPassword)`. The
+`ApplicationPassword` type implements `PartialEq` so two application passwords
+are equal if their label is equal and they refer to the same application.
+
+(NOTE: We could consider extending the `CredentialType` to add
+`CredentialType::ApplicationPassword<Map<Uuid, ApplicationPassword>>`. Then we
+wouldn't need `Attribute::ApplicationPassword`, nor `Value::ApplicationPassword`,
+neither `ValueSetApplicationPassword`).
+
+The user must be able to delete credentials individually. The generated password
+is only displayed once when the user creates it and it is not possible to
+recover the clear-text form, only hashed form is stored. It is not allowed to
+store duplicated application passwords (same app refer and label).
+
+We do not need temporary locks or holds - users can delete and recreate as
+needed.
+
+### Reference integrity
+
+Since application passwords are related to applications, on delete of an
+application all entries that have a bound application password should be removed
+from user accounts.
+
+### Access controls
+
+The IDM administrators group can manage the applications and user's application
```

Comment:
I think this should be an "application administrator" group. It should be able to define new applications etc. 

The application should also allow "managed by" so that it can have delegated administration.

Finally, "application user password admins" should only be able to assert if a pw is present, or delete them - they should NOT be able to create them on users. Only a user should be able to self-create their own application password.

Reply:
"Admin needs to define an app" rather than "the user defines a label, and uses it" feels like a usability problem, TBH.

Reply:
Well the admin has to create the application first to say "this imap server will use application passwords" then the user who is a memberof the allowed access group may create an application password to use said service. So I don't see what the issue is?

Reply:
clearly I don't remember reading this in full at all, ima try again. ignore me for now 😮‍💨

Reply:
No problem <3